exegol/model/ExegolContainer.py

```diff
@@ -334,16 +338,39 @@ def __applyXhostACL(self):
                 return
-
-            if EnvInfo.isMacHost():
-                logger.debug(f"Adding xhost ACL to localhost")
-                # add xquartz inet ACL
-                with console.status(f"Starting XQuartz...", spinner_style="blue"):
-                    os.system(f"xhost + localhost > /dev/null")
+            
+            # Another way of passing the argument without importing subprocess: os.system('xauth list | grep "${DISPLAY%.*}" > '+f"{self.config.getPrivateVolumePath()}/testyu")
+            logger.debug(f"DISPLAY variable: {GuiUtils.getDisplayEnv()}")
+            tmpXauthority = f"/tmp/.tmpXauthority{random.randint(0, 999)}"
+            output = os.system(f"xauth extract {tmpXauthority} $DISPLAY > /dev/null 2>&1")
+            xauthEntry = subprocess.check_output(f"xauth -f {tmpXauthority} list 2>/dev/null",shell=True).decode()
+            logger.debug(f"xauthEntry to propagate: {xauthEntry}")
+            display_host = GuiUtils.getDisplayEnv().split(':')[0]
+            if display_host=='':
+                logger.debug("Connecting to container from local GUI, no X11 forwarding to set up")
+                # TODO verify that the display format is the same on macOS, otherwise might not set up xauth and xhost correctly 
+                if EnvInfo.isMacHost():
+                    logger.debug(f"Adding xhost ACL to localhost")
+                    # add xquartz inet ACL
+                    with console.status(f"Starting XQuartz...", spinner_style="blue"):
+                        os.system(f"xhost + localhost > /dev/null")
+                else:
+                    logger.debug(f"Adding xhost ACL to local:{self.config.getUsername()}")
+                    # add linux local ACL
+                    os.system(f"xhost +local:{self.config.getUsername()} > /dev/null")
+            elif display_host=="localhost" and self.config.getTextNetworkMode() == "bridge":
+                logger.warning("X11 forwarding won't work on a bridged container unless you specify \"X11UseLocalhost no\" in your host sshd_config")
+            elif display_host=="localhost":
+                logger.debug("X11UseLocalhost directive is set to \"yes\" or unspecified, X11 connection can be received only on loopback");
+                # Modifing the entry to convert <hostname>/unix:<display_number> to localhost:<display_number>
+                xauthEntry = f"localhost:{xauthEntry.split(':')[1]}"
             else:
-                logger.debug(f"Adding xhost ACL to local:{self.config.getUsername()}")
-                # add linux local ACL
-                os.system(f"xhost +local:{self.config.getUsername()} > /dev/null")
+                logger.debug("X11UseLocalhost directive is set to \"no\", X11 connection can be received from anywere");
+            if xauthEntry:
+                logger.debug(f"Adding xauth cookie to container: {xauthEntry}")
+                self.exec(f"xauth add {xauthEntry}", as_daemon=False, quiet=True)
```

Comment:
Is this setup still works with multiple shell ? No side effect ?
(This function is called every time a shell is spawn in the container)

Reply:
So, it does work with multiple shells. The reason is that the `DISPLAY` variable won't be the same if multiple users SSH on the host. It will increment (`localhost:10`, `localhost:11` etc.) and add multiple cookies to the `.Xauthority` file of the container. This command is not executed every time the shell spawns but only when the host has a xauth entry corresponding to the current `DISPLAY` variable. Moreover, I used a random to decrease the chance of file collision. Finally, you made a good point saying that I should return and initialize latter the temporary `.Xauthority` file so it will get executed even less.